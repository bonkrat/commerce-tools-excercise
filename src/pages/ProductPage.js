import React from "react";
import { useParams, useHistory } from "react-router-dom";
import { useQuery } from "@apollo/react-hooks";
import { gql } from "apollo-boost";
import LinearProgress from "@material/react-linear-progress";
import moment from "moment";

export const PRODUCT_QUERY = gql`
  query getProduct($id: String!) {
    product(id: $id) {
      id
      createdAt
      masterData {
        current {
          name(locale: "en")
          slug(locale: "en")
          categories {
            name(locale: "en")
          }
        }
      }
    }
  }
`;

/**
 * Creates a view with a product card.
 */
export default () => {
  const history = useHistory();
  const { productId: id } = useParams();
  const { loading, error, data } = useQuery(PRODUCT_QUERY, {
    variables: { id }
  });

  // Redirect if there is an error
  error && history.push("/");

  return (
    <div className="container mx-auto mb-6 w-1/2">
      <LinearProgress
        indeterminate
        className={loading ? "opacity-100" : "opacity-0"}
      />
      <button
        class="mb-4 w-full bg-blue-300 hover:bg-blue-400 text-gray-100 font-bold py-2 px-4 rounded-l"
        onClick={() => history.push("/")}
      >
        Back to Product Search
      </button>
      {!loading && (
        <div class="rounded overflow-hidden shadow-lg min-h-16">
          <img
            class="w-full bg-gray-100"
            src="https://placedog.net/250?random"
            alt="Product Placeholder"
          />
          <div class="px-6 py-4">
            <div class="font-bold text-xl mb-2">
              {data?.product?.masterData?.current?.name}
            </div>
            <div class="font-italics text-s mb-2">
              {data?.product?.masterData?.current?.slug}
            </div>
            <div class="font-italics text-s mb-2">
              {data?.product?.createdAt &&
                moment
                  .utc(data.product.createdAt)
                  .format("MMMM Do YYYY, h:mm:ss a")}
            </div>
          </div>
          <div class="px-6 py-4">
            {data?.product?.masterData?.current?.categories?.map(category => (
              <span
                key={category.name}
                class="inline-block bg-gray-200 rounded-full px-3 py-1 text-sm font-semibold text-gray-700 mr-2"
              >
                #{category?.name}
              </span>
            ))}
          </div>
        </div>
      )}
    </div>
  );
};
