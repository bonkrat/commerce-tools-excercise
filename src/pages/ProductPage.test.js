import React from "react";
import ProductPage, { PRODUCT_QUERY } from "./ProductPage";
import { shallow } from "enzyme";
import { useHistory } from "react-router-dom";
import { useQuery } from "@apollo/react-hooks";
import LinearProgress from "@material/react-linear-progress";

jest.mock("react-router-dom", () => ({
  useParams: jest.fn(() => ({ productId: "foo" })),
  useHistory: jest.fn(() => ({
    push: jest.fn()
  }))
}));

jest.mock("@apollo/react-hooks", () => ({
  useQuery: jest.fn(() => ({
    loading: true,
    error: false,
    data: []
  }))
}));

describe("ProductPage", () => {
  afterEach(() => {
    useQuery.mockReset();
  });
  shallow(<ProductPage />);
  it("queries for a products data", () => {
    expect(useQuery).toHaveBeenCalledWith(PRODUCT_QUERY, {
      variables: { id: "foo" }
    });
  });
  it("redirects if there is an error while querying for data", () => {
    expect(useHistory).toHaveBeenCalled();
    useQuery.mockImplementation(() => ({
      loading: false,
      error: true,
      data: []
    }));
    shallow(<ProductPage />);
    expect(useHistory.mock.results[1].value.push.mock.calls[0]).toEqual(["/"]);
  });
  it("shows a progress indicator while loading", () => {
    useQuery.mockImplementation(() => ({
      loading: true,
      error: false,
      data: []
    }));
    const wrapper = shallow(<ProductPage />);
    expect(wrapper.find(LinearProgress).prop("className")).toEqual(
      "opacity-100"
    );
  });
  it("renders a back button to the product list", () => {
    useQuery.mockImplementation(() => ({
      loading: false,
      error: false,
      data: []
    }));

    const wrapper = shallow(<ProductPage />).findWhere(
      n => n.text() === "Back to Product Search" && n.type() === "button"
    );
    expect(wrapper.exists()).toBeTruthy();
    wrapper.prop("onClick")();
    expect(useHistory.mock.results[1].value.push.mock.calls[0]).toEqual(["/"]);
  });

  [
    ["name", "fooname"],
    ["slug", "fooslug"],
    ["date", "February 9th 2020, 7:03:57 pm"],
    ["tag", "#foocategory"]
  ].forEach(([name, text]) => {
    it(`renders the ${name} value for the product`, () => {
      useQuery.mockImplementation(() => ({
        loading: false,
        error: false,
        data: {
          product: {
            createdAt: 1581275037807,
            masterData: {
              current: {
                name: "fooname",
                slug: "fooslug",
                categories: [
                  {
                    name: "foocategory"
                  }
                ]
              }
            }
          }
        }
      }));

      const wrapper = shallow(<ProductPage />);
      expect(wrapper.findWhere(n => n.text() === text).exists()).toBeTruthy();
    });
  });
});

describe("PRODUCT_QUERY", () => {
  it("queries for product data", () => {
    expect(PRODUCT_QUERY).toMatchSnapshot();
  });
});
