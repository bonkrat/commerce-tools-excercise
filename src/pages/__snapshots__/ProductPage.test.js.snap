// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PRODUCT_QUERY queries for product data 1`] = `
Object {
  "definitions": Array [
    Object {
      "directives": Array [],
      "kind": "OperationDefinition",
      "name": Object {
        "kind": "Name",
        "value": "getProduct",
      },
      "operation": "query",
      "selectionSet": Object {
        "kind": "SelectionSet",
        "selections": Array [
          Object {
            "alias": undefined,
            "arguments": Array [
              Object {
                "kind": "Argument",
                "name": Object {
                  "kind": "Name",
                  "value": "id",
                },
                "value": Object {
                  "kind": "Variable",
                  "name": Object {
                    "kind": "Name",
                    "value": "id",
                  },
                },
              },
            ],
            "directives": Array [],
            "kind": "Field",
            "name": Object {
              "kind": "Name",
              "value": "product",
            },
            "selectionSet": Object {
              "kind": "SelectionSet",
              "selections": Array [
                Object {
                  "alias": undefined,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "id",
                  },
                  "selectionSet": undefined,
                },
                Object {
                  "alias": undefined,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "createdAt",
                  },
                  "selectionSet": undefined,
                },
                Object {
                  "alias": undefined,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "masterData",
                  },
                  "selectionSet": Object {
                    "kind": "SelectionSet",
                    "selections": Array [
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "current",
                        },
                        "selectionSet": Object {
                          "kind": "SelectionSet",
                          "selections": Array [
                            Object {
                              "alias": undefined,
                              "arguments": Array [
                                Object {
                                  "kind": "Argument",
                                  "name": Object {
                                    "kind": "Name",
                                    "value": "locale",
                                  },
                                  "value": Object {
                                    "block": false,
                                    "kind": "StringValue",
                                    "value": "en",
                                  },
                                },
                              ],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "name",
                              },
                              "selectionSet": undefined,
                            },
                            Object {
                              "alias": undefined,
                              "arguments": Array [
                                Object {
                                  "kind": "Argument",
                                  "name": Object {
                                    "kind": "Name",
                                    "value": "locale",
                                  },
                                  "value": Object {
                                    "block": false,
                                    "kind": "StringValue",
                                    "value": "en",
                                  },
                                },
                              ],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "slug",
                              },
                              "selectionSet": undefined,
                            },
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "categories",
                              },
                              "selectionSet": Object {
                                "kind": "SelectionSet",
                                "selections": Array [
                                  Object {
                                    "alias": undefined,
                                    "arguments": Array [
                                      Object {
                                        "kind": "Argument",
                                        "name": Object {
                                          "kind": "Name",
                                          "value": "locale",
                                        },
                                        "value": Object {
                                          "block": false,
                                          "kind": "StringValue",
                                          "value": "en",
                                        },
                                      },
                                    ],
                                    "directives": Array [],
                                    "kind": "Field",
                                    "name": Object {
                                      "kind": "Name",
                                      "value": "name",
                                    },
                                    "selectionSet": undefined,
                                  },
                                ],
                              },
                            },
                          ],
                        },
                      },
                    ],
                  },
                },
              ],
            },
          },
        ],
      },
      "variableDefinitions": Array [
        Object {
          "defaultValue": undefined,
          "directives": Array [],
          "kind": "VariableDefinition",
          "type": Object {
            "kind": "NonNullType",
            "type": Object {
              "kind": "NamedType",
              "name": Object {
                "kind": "Name",
                "value": "String",
              },
            },
          },
          "variable": Object {
            "kind": "Variable",
            "name": Object {
              "kind": "Name",
              "value": "id",
            },
          },
        },
      ],
    },
  ],
  "kind": "Document",
  "loc": Object {
    "end": 273,
    "start": 0,
  },
}
`;
